File: src/enve_triangleground.cc
/*
(***********************************************************************)
(*                                                                     *)
(* The ENVE project                                                    *)
(*                                                                     *)
(* Copyright (c) 2020-2021, Davide Stocco and Enrico Bertolazzi.       *)
(*                                                                     *)
(* The ENVE project and its components are supplied under the terms of *)
(* the open source BSD 2-Clause License. The contents of the ENVE      *)
(* project and its components may not be copied or disclosed except in *)
(* accordance with the terms of the BSD 2-Clause License.              *)
(*                                                                     *)
(* URL: https://opensource.org/licenses/BSD-2-Clause                   *)
(*                                                                     *)
(*    Davide Stocco                                                    *)
(*    Department of Industrial Engineering                             *)
(*    University of Trento                                             *)
(*    e-mail: davide.stocco@unitn.it                                   *)
(*                                                                     *)
(*    Enrico Bertolazzi                                                *)
(*    Department of Industrial Engineering                             *)
(*    University of Trento                                             *)
(*    e-mail: enrico.bertolazzi@unitn.it                               *)
(*                                                                     *)
(***********************************************************************)
*/

///
/// file: enve_triangleground.cc
///

#include "enve_triangleground.hh"

namespace enve
{

  /*\
   |   _        _                   _                          _
   |  | |_ _ __(_) __ _ _ __   __ _| | ___ _ __ ___   __ _  __| |
   |  | __| '__| |/ _` | '_ \ / _` | |/ _ \ '__/ _ \ / _` |/ _` |
   |  | |_| |  | | (_| | | | | (_| | |  __/ | | (_) | (_| | (_| |
   |   \__|_|  |_|\__,_|_| |_|\__, |_|\___|_|  \___/ \__,_|\__,_|
   |                          |___/
  \*/

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  triangleground::~triangleground(void)
  {
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  triangleground::triangleground(void)
    : triangle(NAN_TRIANGLE),
      m_friction(QUIET_NAN),
      m_normal(NAN_VEC3),
      m_bbox(NAN_AABB)
  {
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  triangleground::triangleground(
    real vertex0_x,
    real vertex0_y,
    real vertex0_z,
    real vertex1_x,
    real vertex1_y,
    real vertex1_z,
    real vertex2_x,
    real vertex2_y,
    real vertex2_z,
    real friction)
    : triangle(vertex0_x, vertex0_y, vertex0_z, vertex1_x, vertex1_y, vertex1_z, vertex2_x, vertex2_y, vertex2_z),
      m_friction(friction)
  {
    this->updateNormal();
    this->m_bbox.clamp(
      point(vertex0_x, vertex0_y, vertex0_z),
      point(vertex1_x, vertex1_y, vertex1_z),
      point(vertex2_x, vertex2_y, vertex2_z));
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  triangleground::triangleground(
    point const &point0,
    point const &point1,
    point const &point2,
    real         friction)
    : triangle(point0, point1, point2),
      m_friction(friction)
  {
    this->updateNormal();
    this->m_bbox.clamp(point0, point1, point2);
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  triangleground::triangleground(
    point const vertex[3],
    real        friction)
    : triangle(vertex),
      m_friction(friction)
  {
    this->updateNormal();
    this->m_bbox.clamp(vertex);
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  triangleground::copy(
    triangleground const &triangleground_in)
  {
    triangle::operator=(triangleground_in);
    this->m_friction  = triangleground_in.m_friction;
    this->m_normal    = triangleground_in.m_normal;
    this->m_bbox      = triangleground_in.m_bbox;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  bool
  triangleground::isApprox(
    triangleground const &triangleground_in,
    real                  tolerance)
    const
  {
    return this->vertex(0).isApprox(triangleground_in.vertex(0)) &&
           this->vertex(1).isApprox(triangleground_in.vertex(1)) &&
           this->vertex(2).isApprox(triangleground_in.vertex(2)) &&
           acme::isApprox(this->m_friction, triangleground_in.m_friction, tolerance);
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  real &
  triangleground::friction(void)
  {
    return this->m_friction;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  real const &
  triangleground::friction(void)
    const
  {
    return this->m_friction;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  aabb const &
  triangleground::bbox(void)
    const
  {
    return this->m_bbox;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  vec3 const &
  triangleground::normal(void)
    const
  {
    return this->m_normal;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  void
  triangleground::updateNormal(void)
  {
    this->m_normal = ((this->vertex(1) - this->vertex(0)).cross(this->vertex(2) - this->vertex(0))).normalized();
    if (this->m_normal.z() < 0.0)
      this->m_normal *= -1;
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  point const &
  triangleground::operator[](
    size_t i)
    const
  {
    return triangle::vertex(i);
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  point const &
  triangleground::vertex(
    size_t i)
    const
  {
    return triangle::vertex(i);
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  point &
  triangleground::operator[](
    size_t i)
  {
    return triangle::vertex(i);
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

  point &
  triangleground::vertex(
    size_t i)
  {
    return triangle::vertex(i);
  }

  // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

} // namespace enve

///
/// file: enve_triangleground.cc
///

File: toolbox/lib/enve_triangleground.m
% % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % %
%                                                                     %
% This file is part of the ENVE project.                              %
%                                                                     %
% Copyright (c) 2020, Davide Stocco, Matteo Larcher and Enrico        %
% Bertolazzi.                                                         %
%                                                                     %
% The ENVE project and its components are supplied under the terms of %
% the open source BSD 3-Clause License. The contents of the ENVE      %
% project and its components may not be copied or disclosed except in %
% accordance with the terms of the BSD 3-Clause License.              %
%                                                                     %
%    Davide Stocco                                                    %
%    Department of Industrial Engineering                             %
%    University of Trento                                             %
%    e-mail: davide.stocco@unitn.it                                   %
%                                                                     %
%    Enrico Bertolazzi                                                %
%    Department of Industrial Engineering                             %
%    University of Trento                                             %
%    e-mail: enrico.bertolazzi@unitn.it                               %
%                                                                     %
% % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % %

classdef enve_triangleground < acme_entity
  %>
  %> Class container for ENVE triangleground object
  %>
  methods
    function this = enve_triangleground( varargin )
      %> Create a new C++ pointer to ENVE triangleground object instance
      this.objectHandle = mex_triangleground( 'new', varargin{:} );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Create C++ pointer to ENVE triangleground object instance
    function delete( this )
      mex_triangleground( 'delete', this.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Get ID
    function out = getID( this )
      out = mex_triangleground( 'getID', this.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Get friction coefficent scaling factor
    function out = getFriction( this )
      out = mex_triangleground( 'getFriction', this.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Get triangle vertex 1 as ACME point object
    function out = getVertex1( this )
      out = acme_point();
      out.copyByHandle( mex_triangleground( 'getVertex1', this.objectHandle ) );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Get triangle vertex 2 as ACME point object
    function out = getVertex2( this )
      out = acme_point();
      out.copyByHandle( mex_triangleground( 'getVertex2', this.objectHandle ) );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Get triangle vertex 3 as ACME point object
    function out = getVertex3( this )
      out = acme_point();
      out.copyByHandle( mex_triangleground( 'getVertex3', this.objectHandle ) );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Set triangle vertex 1 with an ACME point object
    function setVertex1( this, other_obj )
      mex_triangleground( 'setVertex1', this.objectHandle, other_obj.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Set triangle vertex 2 with an ACME point object
    function setVertex2( this, other_obj )
      mex_triangleground( 'setVertex2', this.objectHandle, other_obj.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Set triangle vertex 3 with an ACME point object
    function setVertex3( this, other_obj )
      mex_triangleground( 'setVertex3', this.objectHandle, other_obj.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Copy triangleground object from another triangleground
    function copy( this, other_obj )
      if (other_obj.type() == "triangleground")
        mex_triangleground( 'copy', this.objectHandle, other_obj.objectHandle );
      else
        error('mex_triangleground::copy(): other_obj must be an ENVE triangleground object type.');
      end
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Copy triangleground object from another triangleground shape
    function copyByHandle( this, handle )
      mex_triangleground( 'copy', this.objectHandle, handle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Get normal
    function out = normal( this )
      out = mex_triangleground( 'normal', this.objectHandle );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Display object data
    function disp( this )
      disp([[this.getID(); this.getFriction(); NaN], ...
             this.getVertex1().get(), this.getVertex2().get(),  ...
             this.getVertex3().get() ...
           ]);
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Set friction coefficent scaling factor
    function out = toTriangle( this )
      out = acme_triangle( ...
          this.getVertex1().get(), ...
          this.getVertex2().get(), ...
          this.getVertex3().get() ...
        );
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Plot triangle object
    function plot( this, figure_name, color )
      figure_name; %#ok<VUNUS>
      hold on;
      Vertex1 = this.getVertex1().get();
      Vertex2 = this.getVertex2().get();
      Vertex3 = this.getVertex3().get();
      X = [Vertex1(1), Vertex2(1), Vertex3(1)];
      Y = [Vertex1(2), Vertex2(2), Vertex3(2)];
      Z = [Vertex1(3), Vertex2(3), Vertex3(3)];
      patch(X, Y, Z, color, 'FaceAlpha', 1.0);
      hold off;
    end
    %
    % - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    %
    %> Get object type as string
    function out = type( ~ )
      out = 'triangleground';
    end
  end
end
